src/StructureMap.DynamicInterception/MethodInvocation.cs

```diff
@@ -10,28 +10,43 @@ namespace StructureMap.DynamicInterception
     internal class MethodInvocation : ISyncMethodInvocation, IAsyncMethodInvocation
     {
         private readonly IInvocation _invocation;
-        private readonly IList<IArgument> _arguments;
-        private readonly IDictionary<string, IArgument> _argumentMap;
+        private readonly Lazy<List<IArgument>> _arguments;
+        private readonly Lazy<IDictionary<string, IArgument>> _argumentMap;
 
         public MethodInvocation(IInvocation invocation)
         {
             _invocation = invocation;
-            _arguments = invocation.Method.GetParameters()
-                .Zip(invocation.Arguments, (info, value) => new { info, value })
-                .Select((t, i) => new Argument(invocation, i, t.value, t.info))
-                .ToList<IArgument>();
-            _argumentMap = _arguments.ToDictionary(a => a.ParameterInfo.Name);
+            _arguments = new Lazy<List<IArgument>>(() => invocation.Method.GetParameters()
+                                                                    .Zip(invocation.Arguments, (info, value) => new { info, value })
+                                                                    .Select((t, i) => new Argument(invocation, i, t.value, t.info))
+                                                                    .ToList<IArgument>());
+            _argumentMap = new Lazy<IDictionary<string, IArgument>>(() => _arguments.Value.ToDictionary(a => a.ParameterInfo.Name));
         }
 
         public IList<IArgument> Arguments
         {
-            get { return new List<IArgument>(_arguments); }
+            get { return _arguments.Value.AsReadOnly(); }
         }
 
         public IArgument GetArgument(string name)
         {
             IArgument result;
-            return _argumentMap.TryGetValue(name, out result) ? result : null;
+            return _argumentMap.Value.TryGetValue(name, out result) ? result : null;
+        }
+
+        public IArgument GetArgument(int index)
+        {
+            return _arguments.Value[index];
+        }
+
+        public Type[] GenericArguments
+        {
+            get { return _invocation.GenericArguments; }
+        }
+
+        public object InvocationTarget
```

Comment:
IMO I would call it `TargetObject` since it is a part of `MethodInvocation` and `Invocation` part sounds redundant. And then we have `TargetType` property.